RecoJets/JetProducers/plugins/HFJetShowerShape.cc

```diff
@@ -0,0 +1,240 @@
+  vertices_token_ = consumes<std::vector<reco::Vertex>>(iConfig.getParameter<edm::InputTag>("theVertices"));
+  jetPtThreshold_ = iConfig.getParameter<double>("jetPtThreshold");
+  jetEtaThreshold_ = iConfig.getParameter<double>("jetEtaThreshold");
+  hfTowerEtaWidth_ = iConfig.getParameter<double>("hfTowerEtaWidth");
+  hfTowerPhiWidth_ = iConfig.getParameter<double>("hfTowerPhiWidth");
+  vertexRecoEffcy_ = iConfig.getParameter<double>("vertexRecoEffcy");
+  offsetPerPU_ = iConfig.getParameter<double>("offsetPerPU");
+  jetReferenceRadius_ = iConfig.getParameter<double>("jetReferenceRadius");
+  stripPtThreshold_ = iConfig.getParameter<double>("stripPtThreshold");
+  widthPtThreshold_ = iConfig.getParameter<double>("widthPtThreshold");
+
+  produces<edm::ValueMap<float>>("sigmaEtaEta");
+  produces<edm::ValueMap<float>>("sigmaPhiPhi");
+  produces<edm::ValueMap<int>>("centralEtaStripSize");
+  produces<edm::ValueMap<int>>("adjacentEtaStripsSize");
+}
+
+HFJetShowerShape::~HFJetShowerShape() {
+  // do anything here that needs to be done at destruction time
+  // (e.g. close files, deallocate resources etc.)
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void HFJetShowerShape::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+
+  //Jets
+  edm::Handle<edm::View<reco::Jet>> theJets;
+  iEvent.getByToken(jets_token_, theJets);
+
+  //Vertices
+  edm::Handle<std::vector<reco::Vertex>> theVertices;
+  iEvent.getByToken(vertices_token_, theVertices);
+  int nPV = theVertices->size();
+
+  //Products
+  std::vector<float>* v_sigmaEtaEta = new std::vector<float>;
+  std::vector<float>* v_sigmaPhiPhi = new std::vector<float>;
+  std::vector<int>* v_size_CentralEtaStrip = new std::vector<int>;
+  std::vector<int>* v_size_AdjacentEtaStrips = new std::vector<int>;
+
+  //Et offset for HF PF candidates
+  double PUoffset = offsetPerPU_ / (TMath::Pi() * jetReferenceRadius_ * jetReferenceRadius_) * nPV / vertexRecoEffcy_ *
+                    (hfTowerEtaWidth_ * hfTowerPhiWidth_);
+
+  for (auto jet = theJets->begin(); jet != theJets->end(); ++jet) {
+    double pt_jet = jet->pt();
+    double eta_jet = jet->eta();
+
+    //If central or low pt jets, fill with dummy variables
+    if (pt_jet <= jetPtThreshold_ || fabs(eta_jet) <= jetEtaThreshold_) {
+      v_sigmaEtaEta->push_back(-1.);
+      v_sigmaPhiPhi->push_back(-1.);
+      v_size_CentralEtaStrip->push_back(0);
+      v_size_AdjacentEtaStrips->push_back(0);
+    } else {
+      //First loop over PF candidates to compute some global variables needed for shower shape calculations
+      double sumptPFcands = 0.;
+
+      for (unsigned i = 0; i < jet->numberOfSourceCandidatePtrs(); ++i) {
+        reco::CandidatePtr pfJetConstituent = jet->sourceCandidatePtr(i);
+        const reco::Candidate* icand = pfJetConstituent.get();
```

Comment:
```suggestion
        const reco::Candidate* icand = jet->sourceCandidatePtr(i).get();
```
just a bit less verbose